servers/zts/src/main/java/com/yahoo/athenz/zts/store/impl/S3ChangeLogStoreThread.java

```diff
@@ -0,0 +1,337 @@
+            boolean listTruncated = objectListing.isTruncated();
+
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("listObjects: retrieved {} objects, more objects available - {}",
+                        objectSummaries.size(), listTruncated);
+            }
+
+            for (S3ObjectSummary objectSummary : objectSummaries) {
+
+                // if mod time is specified then make sure we automatically skip
+                // any domains older than the specified value
+
+                if (modTime > 0 && objectSummary.getLastModified().getTime() <= modTime) {
+                    continue;
+                }
+
+                // for now skip any folders/objects that start with '.'
+
+                objectName = objectSummary.getKey();
+                if (objectName.charAt(0) == '.') {
+                    continue;
+                }
+                domains.add(objectName);
+            }
+
+            // check if the object listing is truncated or not (break out in this case)
+            // technically we can skip this call and just call listNextBatchOfResults
+            // since that returns null if the object listing is not truncated but
+            // this direct check here makes the logic easier to follow
+
+            if (!listTruncated) {
+                break;
+            }
+
+            objectListing = s3.listNextBatchOfObjects(objectListing);
+        }
+    }
+
+    @Override
+    public List<String> getLocalDomainList() {
+
+        // check to see if we need to maintain our last modification time.
+        // this will be necessary if our last mod time field is null. We need
+        // to save the timestamp at the beginning just in case we end up getting
+        // paged results and while processing the last page, the Syncer pushes
+        // updated domains from the earlier pages
+
+        if (lastModTime == 0) {
+            lastModTime = System.currentTimeMillis();
+        }
+
+        // we are going to initialize our s3 client here since
+        // this is the first entry point before we start
+        // fetching all the domains individually
+
+        awsS3Client = getS3Client();
+
+        ArrayList<String> domains = new ArrayList<>();
+        listObjects(awsS3Client, domains, 0);
+
+        if (!getAllSignedDomains(domains)) {
+            if (!getAllSignedDomains(domains)) {
+                throw new RuntimeException("Interrupted Exception in getAllSignedDomains");
+            }
+        }
+        return domains;
+    }
+
+    private boolean getAllSignedDomains(List<String> domains) {
+        ExecutorService threadPoolExecutor = getExecutorService();
+
+        AmazonS3 tempS3 = getS3Client();
+        for (String domain: domains) {
+            threadPoolExecutor.execute(new ObjectS3Thread(domain, tempSignedDomainMap, tempS3));
+        }
+        threadPoolExecutor.shutdown();
+        while (true) {
```

Comment:
does this mean if for some reason one of the threads is blocked, we'll be blocked forever?

we should probably have some configurable timeout such that if we don't finish within 30 mins (default value), just return as failure and not block forever.